contrib/babelfishpg_tsql/sql/sys_views.sql

```diff
@@ -2102,22 +2098,41 @@ SELECT CAST('TSQL Default TCP' AS sys.sysname) AS name
   , CAST(0 AS sys.bit) AS is_admin_endpoint;
 GRANT SELECT ON sys.endpoints TO PUBLIC;
 
-create or replace view sys.index_columns
-as
-select i.indrelid::integer as object_id
-  , CAST(CASE WHEN i.indisclustered THEN 1 ELSE 1+row_number() OVER(PARTITION BY c.oid) END AS INTEGER) AS index_id
-  , a.attrelid::integer as index_column_id
-  , a.attnum::integer as column_id
-  , a.attnum::sys.tinyint as key_ordinal
-  , 0::sys.tinyint as partition_ordinal
-  , 0::sys.bit as is_descending_key
-  , 1::sys.bit as is_included_column
-from pg_index as i
-inner join pg_catalog.pg_attribute a on i.indexrelid = a.attrelid
-inner join pg_class c on i.indrelid = c.oid
-inner join sys.schemas sch on sch.schema_id = c.relnamespace
-where has_schema_privilege(sch.schema_id, 'USAGE')
-and has_table_privilege(c.oid, 'SELECT,INSERT,UPDATE,DELETE,TRUNCATE,TRIGGER');
+CREATE OR REPLACE VIEW sys.index_columns
+AS
+SELECT
+    CAST(i.indrelid AS INT) AS object_id,
+    -- should match index_id of sys.indexes 
+    CAST(CASE
+            WHEN i.indisclustered THEN 1
+            ELSE (SELECT COUNT(*)+2 FROM pg_index ind WHERE ind.indrelid = i.indrelid AND ind.indexrelid < i.indexrelid AND ind.indisclustered = false)
+         END AS INT) AS index_id,
+    CAST(a.index_column_id AS INT) AS index_column_id,
+    CAST(a.attnum AS INT) AS column_id,
+    CAST(CASE
+            WHEN a.index_column_id <= i.indnkeyatts THEN a.index_column_id
+            ELSE 0
+         END AS SYS.TINYINT) AS key_ordinal,
+    CAST(0 AS SYS.TINYINT) AS partition_ordinal,
+    CAST(CASE
+            WHEN i.indoption[a.index_column_id-1] & 1 = 1 THEN 1
+            ELSE 0 
+         END AS SYS.BIT) AS is_descending_key,
+    CAST(CASE
+            WHEN a.index_column_id > i.indnkeyatts THEN 1
+            ELSE 0
+         END AS SYS.BIT) AS is_included_column
```

Comment:
Please explain these changes in description or as a reply to this comment

Reply:
pg_index. indkey (vector) containing values that indicate which table columns this index indexes.
Key columns are placed first and then included columns.
index_column_id is the iterator of this array. 1,2,3.... 

key ordinal is column_id for index but on key attributes
`a.index_column_id <= i.indnkeyatts` check if iterator is less than total keys of index. otherwise it is not a key column

is_descending_key
for every entry in indkey vector (key columns), hold the acess method information for the key
otherwise zero

is_included_column
we have the number of keys in index and we know index keys are always first in indkey vector.
hence `a.index_column_id > i.indnkeyatts`